src/Core_Command.php

```diff
@@ -1372,7 +1372,69 @@ private function cleanup_extra_files( $version_from, $version_to, $locale, $inse
 			return;
 		}
 
-		$files_to_remove = array_diff( array_keys( $old_checksums ), array_keys( $new_checksums ) );
+		// Compare the files from the old version and the new version in a case-insensitive manner,
+		// to prevent files being incorrectly deleted on systems with case-insensitive filesystems
+		// when core changes the case of filenames.
+		// The main logic for this was taken from the Joomla project and adapted for WP.
+		// See: https://github.com/joomla/joomla-cms/blob/bb5368c7ef9c20270e6e9fcc4b364cd0849082a5/administrator/components/com_admin/script.php#L8158
+
+		$old_filepaths = array_keys( $old_checksums );
+		$new_filepaths = array_keys( $new_checksums );
+
+		$new_filepaths = array_combine( array_map( 'strtolower', $new_filepaths ), $new_filepaths );
```

Comment:
AFAIK if we pass new checksums trough `strtolower` then we lose case information witch are very important in this code.

Reply:
This line builds an array that use the lowercased string as the key and the original string as the value, to help with mapping.

Reply:
Note the `array_combine()`.

So we'll go from this:
```
[
   0 => 'Abc',
   1 => 'BCD',
]
```
to this:
```
[
   'abc' => 'Abc',
   'bcd' => 'BCD',
]
```